### PR TITLE
Add read muon code 

### PR DESCRIPTION
**Pull request by hggvfdg**

### CODE REVIEW

File: find_max_r_and_z.C [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+        double r = sqrt(x*x+y*y);
+        if (r>max_r) max_r = r;
+
+        double fabs_z = fabs(pmt.GetPosition(1));
+        if (fabs_z>max_z) max_z = fabs_z;
+    }
+
+    std::cout << "Maximum radius: " << max_r << "\n" <<
+        "Maximum z: " << max_z << "\n";
+
```
Comment (kmtsui) [inline find_max_r_and_z.C]: As a good practice, close the file with `file->Close();`
Sometimes you might also want to `delete` all pointers declared by `new`, in order to avoid memory leaks.

File: multiple_events_reader.C [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+    int nEvent = t->GetEntries();
+    std::cout << "Total number of event in all the given files: " << nEvent << "\n\n";
+    WCSimRootEvent* wcsimrootsuperevent = new WCSimRootEvent();
+    t->SetBranchAddress("wcsimrootevent", &wcsimrootsuperevent);
+    WCSimRootTrack* wcsimroottrack = new WCSimRootTrack();
+
+    for (int i=0; i<nEvent; i++) {
+
+        //progress
+        std::cout << "\rReading event " << i << "/" << nEvent;
```
Comment (kmtsui) [inline multiple_events_reader.C]: Better not to print every events, maybe do it for e.g. every 1 percent of all events

File: multiple_events_reader.C [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+        //Determine the location of the entrance and exit point
+        int entrance_boundary_index, exit_boundary_index;
+        if (nCrossing<5) {
+            entrance_boundary_index = 1;
+            exit_boundary_index = 2;
+        }
+        else {
+            entrance_boundary_index = 0;
+            exit_boundary_index = 4;
```
Comment (kmtsui) [inline multiple_events_reader.C]: why not the last index, i.e. `exit_boundary_index=nCrossing-1;`

File: multiple_events_reader.C [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+
+        //Ploting total charge vs energy loss diagram
+        double E_loss = wcsimroottrack->GetBoundaryKEs().at(entrance_boundary_index) - wcsimroottrack->GetBoundaryKEs().at(exit_boundary_index);
+        double total_charge = 0;
+        int nhits = wcsimrootevent->GetNcherenkovdigihits();
+        for (int i=0; i<nhits; i++) {
+            WCSimRootCherenkovDigiHit* wcsimrootcherenkovdigihit = (WCSimRootCherenkovDigiHit*) (wcsimrootevent->GetCherenkovDigiHits())->At(i);
+            total_charge += wcsimrootcherenkovdigihit->GetQ();
+        }
+        Q_vs_E->Fill(E_loss, total_charge, 1);
```
Comment (kmtsui) [inline multiple_events_reader.C]: simply `Q_vs_E->Fill(E_loss, total_charge);` is good

File: read_main_track.C [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+    std::cout << "Total number of events: " << nEvent << "\n";
+    if (iEvent >= nEvent) {
+        std::cout << "Event " << iEvent << " does not exist\n";
+        return 1;
+    }
+
+    //Construct the event branch
+    WCSimRootEvent* wcsimrootsuperevent = new WCSimRootEvent();
+    tree->SetBranchAddress("wcsimrootevent",&wcsimrootsuperevent);
+    tree->GetBranch("wcsimrootevent")->SetAutoDelete(kTRUE);    //Idk what is this for but things seem to be ok without it.
```
Comment (kmtsui) [inline read_main_track.C]: if you only read a single event, you don't need this line. However, when reading multiple event, this is needed to avoid memory leaks

File: multiple_events_reader.C [hunk 6]
```diff
@@ -1,6 +1,6 @@
-//max_r and max_z are obtained from find_max_r_and_z.C
-#define max_r 156.763
-#define max_z 137.928
+//max_r and max_z is taken to be the outest boundary of the tank
+#define max_r 162.67
+#define max_z 143.542
```
Comment (kmtsui) [inline multiple_events_reader.C]: Any instructions on how to get these values in the first place?

File: WCSim_fitQun.c [hunk 7]
```diff
@@ -0,0 +1,278 @@
...
+        exit[2] = z+z_dir*vt;
+        //check if the particle passes through the barrel
+        if ((exit[0]*exit[0]+exit[1]*exit[1])>(R*R)) {
+            double cos_theta = -(x_dir*x+y_dir*y)/(r_dir*r);
+            vt = (r/r_dir)*(cos_theta+sqrt(cos_theta*cos_theta+(R*R)/(r*r)-1));
+            exit[0] = x+x_dir*vt;
+            exit[1] = y+y_dir*vt;
+            exit[2] = z+z_dir*vt;
+        }
+        //search for entrance point
```
Comment (kmtsui) [inline WCSim_fitQun.c]: the code for entrance and exit points are fundamentally the same, try to make a function that can be used in both situations

File: WCSim_fitQun.c [hunk 8]
```diff
@@ -0,0 +1,278 @@
...
+            vt = (r/r_dir)*(cos_theta+sqrt(cos_theta*cos_theta+(R*R)/(r*r)-1));
+            entrance[0] = x+x_dir*vt;
+            entrance[1] = y+y_dir*vt;
+            entrance[2] = z+z_dir*vt;
+        }
+        return false;
+    }
+
+    //check whether the vertex position is on top/at the bottom of the tank, and perform forward extrapolation
+    if (z_dir>0) {vt = (-Z-z)/z_dir;}
```
Comment (kmtsui) [inline WCSim_fitQun.c]: again the code looks very similar as above, try to make a function to reuse the code

File: m_selection.c [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+        percentage->Draw("hist");
+        c1->SaveAs(Form("percentage.pdf"));
+        delete percentage;
+    }
+
+    static bool selector() {
+        //if (fitQunLikelihood>=cut) {return true;}
+        if (fitQunEntrance[2]==max_z) {return true;}
+        if (fitQunExit[2]==-max_z) {return true;}
+        if (fitQunDist<=cut) {return true;}
```
Comment (kmtsui) [inline m_selection.c]: I guess you need to comment and uncomment these lines every time when you change the cut variables? Try to make this function to take an argument to identity the cut conditions.

File: m_selection.c [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+        dist_diff_sd->Draw("hist");
+        c1->SaveAs(Form("dist_diff_sd.pdf"));
+        delete dist_diff_sd;
+        percentage->Draw("hist");
+        c1->SaveAs(Form("percentage.pdf"));
+        delete percentage;
+    }
+
+    static bool selector() {
+        //if (fitQunLikelihood>=cut) {return true;}
```
Comment (kmtsui) [inline m_selection.c:177]: Should it be `fitQunLikelihood<=cut` if we want to select good quality fit?

File: m_selection.c [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+    data.push_back(entrance_diff->GetMean());
+    data.push_back(entrance_diff->GetStdDev());
+    delete entrance_diff;
+    data.push_back(exit_diff->GetMean());
+    data.push_back(exit_diff->GetStdDev());
+    delete exit_diff;
+    data.push_back(dist_diff->GetMean());
+    data.push_back(dist_diff->GetStdDev());
+    delete dist_diff;
+    data.push_back(((double)(nEntries-ignored_track))/1000);
```
Comment (kmtsui) [inline m_selection.c]: why divided by 1000 instead of nEntries?

File: s_selection.c [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+//max_r and max_z is taken to be the outest boundary of the tank
+#define max_r 162.67
+#define max_z 143.542
+
+//function that perform selection
+//Selection option: 1 top to bottom, 2 top to bottom with cutoff, 3 top to barrel, 4 barrel to bottom, 5 barrel to barrel
+bool selector(const int SelectOpt, const double fitQunPos[3], const double fitQunEntrance[3], const double fitQunExit[3], const double EntranceCut = 120, const double ExitCut = 145) {
+
+    if (SelectOpt==1) {
+        if (fitQunEntrance[2]!=max_z) {return true;}        //selecting the fitQun entrance point on the top of the tank
```
Comment (kmtsui) [inline s_selection.c]: better not use `==` or `!=` for floating point comparison

File: AngAcceptance.c [hunk 13]
```diff
@@ -0,0 +1,217 @@
...
+                (fitQunEntrance[1]-fitQunPos[1])*(fitQunEntrance[1]-fitQunPos[1]) +
+                (fitQunEntrance[2]-fitQunPos[2])*(fitQunEntrance[2]-fitQunPos[2]))/TMath::Ccgs();
+        }
+        else {
+            fitQunEntranceTime = fitQunVerTime +
+                (1e9)*sqrt((fitQunEntrance[0]-fitQunPos[0])*(fitQunEntrance[0]-fitQunPos[0]) +
+                (fitQunEntrance[1]-fitQunPos[1])*(fitQunEntrance[1]-fitQunPos[1]) +
+                (fitQunEntrance[2]-fitQunPos[2])*(fitQunEntrance[2]-fitQunPos[2]))/TMath::Ccgs();
+        }
+        if (fitQunEntranceTime<0) {
```
Comment (kmtsui) [inline AngAcceptance.c]: Why is this necessary? In real data we don't have an absolute t=0 as a reference point

Comment (hggvfdg) [reply AngAcceptance.c]: The time of fitQun vertex should be agreed with t=0 at the start of the simulation and I am constructing the track of the emitted photon stating form the entrance point. Therefore I add this for a better hit time calculation later.

Comment (kmtsui) [reply AngAcceptance.c]: This is only true for simulation, but we don't know the proper timing for real data. If we want events with better timing resolution, we should look at other cuts but not on the absolute time which is only known for simulation

File: function.h [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+        VerToPMTDir[2] = PMTPos[i][2] - VerPos[2];
+        VerToPMTDist = sqrt(VerToPMTDir[0]*VerToPMTDir[0] + VerToPMTDir[1]*VerToPMTDir[1] + VerToPMTDir[2]*VerToPMTDir[2]);
+        VerToPMTDir[0] = VerToPMTDir[0]/VerToPMTDist;
+        VerToPMTDir[1] = VerToPMTDir[1]/VerToPMTDist;
+        VerToPMTDir[2] = VerToPMTDir[2]/VerToPMTDist;
+
+        //theta is the angle between the vertex direction vector and the VerToPMTDir[3] vector
+        theta = VerToPMTDir[0]*VerDir[0] + VerToPMTDir[1]*VerDir[1] + VerToPMTDir[2]*VerDir[2];     //This is the value of cos(theta)
+        //Reconstruction fail when cos(theta) > 1
+        if (fabs(theta)>1) {
```
Comment (kmtsui) [inline function.h:29]: numerical precision problem? for costh>1, it's better to have costh=1 instead of -1

File: GenerateWCSim.sh [hunk 15]
```diff
@@ -0,0 +1,22 @@
+#!/bin/bash
+
+file_path="./macros/WCTE.mac"
```
Comment (kmtsui) [inline GenerateWCSim.sh]: please also add this file to the repo

File: WCSimfitQunHitPreprocess.c [hunk 16]
```diff
@@ -0,0 +1,368 @@
...
+
+            int FirstHitPos = WRCH->GetTotalPe(0), LastHitPos = WRCH->GetTotalPe(0)+WRCH->GetTotalPe(1)-1;     //The position on the Array of WCSimRootCherenkovHitTime of the first hit and the last hit on this PMT
+            //loop for the true hits of a PMT in an event and fill the THTree
+            for (int k=FirstHitPos; k<=LastHitPos; k++) {
+                WCSimRootCherenkovHitTime* WRCHT = (WCSimRootCherenkovHitTime*) wcsimrootevent->GetCherenkovHitTimes()->At(k);
+                if (WRCHT->GetParentSavedTrackID()==(-1)) {
+                    THNum--;
+                    continue;
+                }   //ignore dark noise
+
```
Comment (kmtsui) [inline WCSimfitQunHitPreprocess.c:320]: We actually don't care whether a true hit generates a digi hit or not, so you don't need the loop below for selection. Just record every true hit.

File: WCSimfitQunHitPreprocess.c [hunk 17]
```diff
@@ -0,0 +1,368 @@
...
+        WCSimStartPos[1] = wcsimroottrack->GetStart(2);    //g
+        WCSimStartPos[2] = wcsimroottrack->GetStart(1);    //g
+        THNum = wcsimrootevent->GetNcherenkovhittimes();
+
+        TClonesArray* WRCDHArray = wcsimrootevent->GetCherenkovDigiHits();    //Array of WCSimRootCherenkovDigiHit pointer
+        DHNum = WRCDHArray->GetEntriesFast();
+        for (int j=0; j<DHNum; j++) {
+            //ignore extremely long digit time (>200 ns)
+            if (((WCSimRootCherenkovDigiHit*) WRCDHArray->At(j))->GetT()>200) {
+                WRCDHArray->RemoveAt(j);
```
Comment (kmtsui) [inline WCSimfitQunHitPreprocess.c:292]: since you don't need this array to do selection for true hits below, just fill the `DHTree` here,no need to remove those entry.

**Comment by kmtsui** [comment]
Remember to add the instructions of how to run your analysis in your README

File: WCSimfitQunHitPreprocess.c (hunk 16 above)
Comment (hggvfdg) [reply WCSimfitQunHitPreprocess.c:320]: The loop below is not for selection of true hits but to find the DHID. That is, if a true hit generate a digit hit, the DHID of this true hit carries the position of corresponding digit hit in the digit hit tree. This provides a method to relates the true and digit hit tree, though I didn't use this a lot.

File: function.h (hunk 14 above)
Comment (hggvfdg) [reply function.h:29]: you mean manually setting costheta=1 if the value is slightly larger than 1 due to numerical precision problem?

File: WCSimfitQunHitPreprocess.c (hunk 16 above)
Comment (hggvfdg) [reply WCSimfitQunHitPreprocess.c:320]: Maybe I should remove the DHID method because the loop makes the code extremely slow.

File: function.h (hunk 14 above)
Comment (kmtsui) [reply function.h:29]: yes it is
